### PR TITLE
feat: resolve `<include-fragment>` tags e.g. in github release pages

### LINKED ISSUE #34164: `spack checksum` support for GitHub URLs is broken
### Steps to reproduce

As requested by @adamjstewart in #34162, open this as separate issue:

When trying to see the different versions available upstream for a package on github, `spack checksum` does not detect GitHub URLs automatically.

Reprocuce e.g. with package `dcm2niix`:
```
$ spack checksum dcm2niix
==> Found 1 version of dcm2niix:
  
  1.0.20210317  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20210317.tar.gz

==> Fetching https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20210317.tar.gz

    version("1.0.20210317", sha256="42fb22458ebfe44036c3d6145dacc6c1dc577ebbb067bedc190ed06f546ee05a")
```
Expected output:
```
» spack checksum dcm2niix    
==> Found 10 versions of dcm2niix:
  
  1.0.20220720  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20220720.tar.gz
  1.0.20211006  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20211006.tar.gz
  1.0.20210317  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20210317.tar.gz
  1.0.20201102  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20201102.tar.gz
  1.0.20200331  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20200331.tar.gz
  1.0.20190902  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20190902.tar.gz
  1.0.20190720  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20190720.tar.gz
  1.0.20181125  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20181125.tar.gz
  1.0.20181114  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20181114.tar.gz
  1.0.20180622  https://github.com/rordenlab/dcm2niix/archive/refs/tags/v1.0.20180622.tar.gz
...
```

### Error message

_No response_

### Information on your system

* **Spack:** 0.20.0.dev0 (066ec316047cb6dfc92a3fef9980718ade47d4aa)
* **Python:** 3.11.0
* **Platform:** linux-fedora37-skylake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by wdconinc**
This aims to resolve #34164 by resolving the `<include-fragment>` tags that GitHub has started using for their release pages, see https://github.github.io/include-fragment-element/.

This feels a bit hacky but intended as a starting point for discussion. After reading a page during spidering, it first parses for include-fragments, gets them all, and treats them all as separate pages. Then it looks for href links in both the page itself and the fragments.

### CODE REVIEW

**Comment by wdconinc** [comment]
It does find the newer version of `vc`:
```console
$ spack versions --new vc
==> New remote versions (not yet checksummed):
  1.4.3
```

**Comment by wdconinc** [comment]
@alecbcs Can you try this on the `abyss` package you were referring to? This is what I get:
```console
$ spack versions --new abyss
==> New remote versions (not yet checksummed):
  2.3.5  2.3.4  2.3.3  2.3.2
```

File: lib/spack/spack/util/web.py
Comment (wdconinc) [inline lib/spack/spack/util/web.py]: This is failing style checks due to the long line, but before fixing that this needs to be a narrower exception catch in any case. The catch is intended to guard against dead fragment links. It seems release pages (e.g. https://github.com/VcDevel/Vc/releases) include an `include-fragment` with `src` to `/security/overall-count` (e.g. https://github.com/VcDevel/Vc/security/overall-count), which doesn't exist. That causes the code to jump to the outer exception catch without parsing any links ever.

Comment (alecbcs) [reply lib/spack/spack/util/web.py]:
```suggestion
                except Exception as e:
                    msg = f"Error reading fragment: {(type(e), str(e)}:{traceback.format_exc()}"
                    tty.debug(msg)
```

What about converting to an F-string to make `black` happier without splitting up the debug call onto multiple lines?

**Review by alecbcs — COMMENTED** [review]
> @alecbcs Can you try this on the `abyss` package you were referring to? This is what I get:
> 
> ```
> $ spack versions --new abyss
> ==> New remote versions (not yet checksummed):
>   2.3.5  2.3.4  2.3.3  2.3.2
> ```

Nice! Works for me too!

File: lib/spack/spack/util/web.py
Comment (wdconinc) [reply lib/spack/spack/util/web.py]: Do we allow those now? I guess they are since they were introduced in python 3.6 and spack requires that as a minimum.

Comment (alecbcs) [reply lib/spack/spack/util/web.py]: We do! Exactly, now that the minimum python version has moved up they're fair game as I understand it.

**Comment by alecbcs** [comment]
Looks like there's one last style error that'd be pretty easy to resolve. @wdconinc are you okay breaking up that long line as suggested by the black style check output?

**Comment by wdconinc** [comment]
@spackbot fix style

**Comment by wdconinc** [comment]
Now we wait for https://github.com/spack/spack/pull/36757 to fix the remaining CI issues.

**Comment by wdconinc** [comment]
~~https://github.com/spack/spack/pull/36757 is now what's happening.~~ (previous message originally didn't post)

**Comment by wdconinc** [comment]
@spackbot fix style

**Comment by alecbcs** [comment]
@wdconinc looks like Harmen got those unit test fixes in. Can you rebase and and we'll see if everything passes?

**Comment by wdconinc** [comment]
I'm a bit surprised by the codecov diff failure. It seems to think only 11% of the diff in web.py was tested, even after adding the unit test. Anyone know why?

File: lib/spack/spack/util/web.py
Comment (alalazo) [inline lib/spack/spack/util/web.py]: Why calling this function explicitly instead of using:
```suggestion
        super().__init__()
```
?

Comment (wdconinc) [reply lib/spack/spack/util/web.py]: Only because it is done this way in LinkParser, ~10 lines above. I can change both if super is preferred.

Comment (alalazo) [reply lib/spack/spack/util/web.py]: I would change it. `super().__init__()` follows any refactor we might do later, without the possibility of leaving a wrong method call if we overlook that line.

**Review by alecbcs — APPROVED** [review]
This is looking good to me! Thanks @wdconinc for putting in the work to add unit tests. I'll defer to @alalazo to also make sure he's happy with this before we merge.

**Review by alalazo — APPROVED** [review]

**Pull request merged**
